src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+
+	appendStringInfo(
+		query,
+		" SELECT cs.shardid,pd.shardminvalue,pd.shardmaxvalue,cs.shard_size,pn.nodeid,ct.distribution_column,ct.table_name,cs.shard_name,(SELECT relname FROM pg_class WHERE oid = (ct.table_name::regclass)::oid) "
+		" FROM pg_catalog.pg_dist_shard pd JOIN pg_catalog.citus_shards cs ON pd.shardid = cs.shardid JOIN pg_catalog.pg_dist_node pn ON cs.nodename = pn.nodename AND cs.nodeport= pn.nodeport"
+		" JOIN"
+		" ( select shardid , max_size from (SELECT distinct first_value(shardid) OVER w as shardid, sum(shard_size) OVER (PARTITION BY colocation_id, shardminvalue) as total_sum, max(shard_size) OVER w as max_size"
+		" FROM citus_shards cs JOIN pg_dist_shard ps USING(shardid)"
+		" WINDOW w AS (PARTITION BY colocation_id, shardminvalue ORDER BY shard_size DESC) )as t where total_sum >= %ld )"
+		" AS max_sizes ON cs.shardid=max_sizes.shardid AND cs.shard_size = max_sizes.max_size JOIN citus_tables ct ON cs.table_name = ct.table_name AND pd.shardminvalue <> pd.shardmaxvalue AND pd.shardminvalue <> ''",
+		MaxShardSize*1024
+		);
+
+		ereport(LOG, errmsg("%s", query->data));
+		char *shardSplitMode;
+		Oid shardTransferModeOid = PG_GETARG_OID(0);
+		Datum enumLabelDatum = DirectFunctionCall1(enum_out, shardTransferModeOid);
+		shardSplitMode = DatumGetCString(enumLabelDatum);
+		ereport(LOG,errmsg("%s",shardSplitMode));
+	if (SPI_connect() != SPI_OK_CONNECT)
+	{
+		elog(ERROR, "SPI_connect to the query failed");
+	}
+	if (SPI_exec(query->data, 0) != SPI_OK_SELECT)
+	{
+		elog(ERROR, "SPI_exec for the execution failed");
+	}
+
+	SPITupleTable *tupletable = SPI_tuptable;
+	int rowCount = SPI_processed;
+	bool isnull;
+
+	for (int rowIndex = 0; rowIndex < rowCount; rowIndex++)
+	{
+		ShardInfoData shardinfo;
+		HeapTuple tuple = tupletable->vals[rowIndex];
+
+		Datum shardId = SPI_getbinval(tuple, tupletable->tupdesc, 1, &isnull);
+		shardinfo.shardid = DatumGetInt64(shardId);
+
+		Datum shardSize = SPI_getbinval(tuple, tupletable->tupdesc, 4, &isnull);
+		shardinfo.shardsize = DatumGetInt64(shardSize);
+
+		Datum nodeId = SPI_getbinval(tuple, tupletable->tupdesc, 5, &isnull);
+		shardinfo.nodeid = DatumGetInt32(nodeId);
+
+		char *shardMinVal = SPI_getvalue(tuple, tupletable->tupdesc, 2);
+		shardinfo.shardminvalue = strtoi64(shardMinVal, NULL, 10);
+
+		char *shardMaxVal = SPI_getvalue(tuple, tupletable->tupdesc, 3);
+		shardinfo.shardmaxvalue = strtoi64(shardMaxVal, NULL, 10);
+
+		shardinfo.distributionColumn = SPI_getvalue(tuple, tupletable->tupdesc, 6);
+		shardinfo.tablename = SPI_getvalue(tuple, tupletable->tupdesc, 7);
+
+		StringInfo shardnameQuery = makeStringInfo();
+		appendStringInfo(shardnameQuery, "%s_%ld", SPI_getvalue(tuple,
+																tupletable->tupdesc, 9),
+						 shardinfo.shardid);
+		shardinfo.shardname = shardnameQuery->data;
+
+		Datum tableIdDatum = SPI_getbinval(tuple, tupletable->tupdesc, 7, &isnull);
+		shardinfo.tableId = DatumGetObjectId(tableIdDatum);
```

Comment:
maybe instead of storing the tablename in the ShardInfoData, we just store the tableId and get the name and schema as needed using get_rel_name(tableId) and get_rel_namespace(tableId).

Reply:
removed schemaname however tablename in shardinfo stores tablename appended with schemaname so it is directly getting used in query